cloudify/amqp_client.py

```diff
@@ -263,7 +276,46 @@ def close(self, wait=True):
     def add_handler(self, handler):
         self._handlers.append(handler)
         if self._pika_connection:
-            handler.register(self._pika_connection, self._publish_queue)
+            handler.register(self)
+
+    def channel(self):
+        if self._closed or not self._pika_connection:
+            raise RuntimeError(
+                'Attempted to open a channel on a closed connection')
+        return self._pika_connection.channel()
+
+    def publish(self, message, wait=True):
+        """Schedule a message to be sent.
+
+        :param message: Kwargs for the pika basic_publish call. Should at
+                        least contain the "body" and "exchange" keys, and
+                        it might contain other keys such as "routing_key"
+                        or "properties"
+        :param wait: Whether to wait for the message to actually be sent.
+                     If true, an exception will be raised if the message
+                     cannot be sent.
+        """
+        if wait and self._consumer_thread \
+                and self._consumer_thread is threading.current_thread():
```

Comment:
If this behavior is broken, why is it allowed in the first place?

Reply:
it's not, this line disallows it :P
other than that, we have no way to stop callbacks from calling `.publish`